timemachine/cpp/src/kernels/k_flat_bottom_bond.cuh

```diff
@@ -115,39 +122,48 @@ void __global__ k_flat_bottom_bond(
     }
+    if (du_dp || du_dx) {
+        RealType d_r_min = r - rmin;
+        RealType d_r_max = r - rmax;
+        RealType d_rmin_3 = d_r_min * d_r_min * d_r_min;
+        RealType d_rmax_3 = d_r_max * d_r_max * d_r_max;
+        if (du_dp) {
+            // compute parameter derivatives
+            RealType du_dk_real =
+                (r_gt_rmax * ((d_rmax_3 * d_r_max) / static_cast<RealType>(4)) +
+                 (r_lt_rmin * ((d_rmin_3 * d_r_min) / static_cast<RealType>(4))));
+            RealType du_drmin_real = r_lt_rmin * (-k * d_rmin_3);
+            RealType du_drmax_real = r_gt_rmax * (-k * d_rmax_3);
 
-    if (du_dp) {
-        // compute parameter derivatives
-        RealType du_dk_real = (r_gt_rmax * (pow(r - rmax, 4) / 4)) + (r_lt_rmin * (pow(r - rmin, 4) / 4));
-        RealType du_drmin_real = r_lt_rmin * (-k * pow(r - rmin, 3));
-        RealType du_drmax_real = r_gt_rmax * (-k * pow(r - rmax, 3));
-
-        // cast float -> fixed
-        auto du_dk = FLOAT_TO_FIXED_BONDED<RealType>(du_dk_real);
-        auto du_drmin = FLOAT_TO_FIXED_BONDED<RealType>(du_drmin_real);
-        auto du_drmax = FLOAT_TO_FIXED_BONDED<RealType>(du_drmax_real);
-
-        // increment du_dp array
-        atomicAdd(du_dp + k_idx, du_dk);
-        atomicAdd(du_dp + rmin_idx, du_drmin);
-        atomicAdd(du_dp + rmax_idx, du_drmax);
-    }
+            // cast float -> fixed
+            unsigned long long du_dk = FLOAT_TO_FIXED_BONDED<RealType>(du_dk_real);
+            unsigned long long du_drmin = FLOAT_TO_FIXED_BONDED<RealType>(du_drmin_real);
+            unsigned long long du_drmax = FLOAT_TO_FIXED_BONDED<RealType>(du_drmax_real);
+
+            // increment du_dp array
+            atomicAdd(du_dp + k_idx, du_dk);
+            atomicAdd(du_dp + rmin_idx, du_drmin);
+            atomicAdd(du_dp + rmax_idx, du_drmax);
+        }
 
-    if (du_dx) {
-        RealType du_dr = k * ((r_gt_rmax * pow(r - rmax, 3)) + (r_lt_rmin * pow(r - rmin, 3)));
+        if (du_dx) {
+            RealType du_dr = k * ((r_gt_rmax * d_rmax_3) + (r_lt_rmin * d_rmin_3));
 
-        for (int d = 0; d < 3; d++) {
-            // compute du/dcoords
-            RealType du_dsrc_real = du_dr * dx[d] / r;
-            RealType du_ddst_real = -du_dsrc_real;
+            RealType inv_r = 1 / r;
+#pragma unroll
+            for (int d = 0; d < 3; d++) {
+                // compute du/dcoords
+                RealType du_dsrc_real = du_dr * dx[d] * inv_r;
+                RealType du_ddst_real = -du_dsrc_real;
 
-            // cast float -> fixed
-            auto du_dsrc = FLOAT_TO_FIXED_BONDED<RealType>(du_dsrc_real);
-            auto du_ddst = FLOAT_TO_FIXED_BONDED<RealType>(du_ddst_real);
+                // cast float -> fixed
+                unsigned long long du_dsrc = FLOAT_TO_FIXED_BONDED<RealType>(du_dsrc_real);
+                unsigned long long du_ddst = FLOAT_TO_FIXED_BONDED<RealType>(du_ddst_real);
```

Comment:
It is implied, and I would prefer we are explicit. We typically avoid using `auto` and don't need it adds anything other than ambiguity.

Reply:
Also, this might just me being unable to read, but I initially did `RealType du_src = FLOAT_TO_FIXED` which does compile and fail in a subtle way. Obvious `auto` does do the right thing here

Reply:
+1 to no `autos`, esp. in performance critical GPU kernels.